lib/pbench/server/cache_manager.py

```diff
@@ -646,6 +665,41 @@ def extract(tarball_path: Path, path: str) -> Inventory:
             tarball_path, f"Unexpected error from {tar_path}: {error_text!r}"
         )
 
+    def stream(self, path: str) -> Inventory:
+        """Return a cached inventory file as a binary stream
+
+        Args:
+            path: Relative path of a regular file within a tarball.
+
+        Returns:
+            An Inventory object encapsulating the file stream and the cache
+            lock. The caller is reponsible for closing the Inventory!
+        """
+        self.cache_create()
+
+        # NOTE: the cache is unlocked here. That's "probably OK" as the reclaim
+        # is based on the last modified timestamp of the lock file and only
+        # runs periodically.
+        #
+        # TODO: Can we convert the lock from EXCLUSIVE to SHARED? If so, that
+        # would be ideal.
```

Comment:
Sadly, it doesn't actually matter. Just before I left this afternoon, I tried to experiment with lock conversion interactively. The mapping of Python file locking APIs to kernel functions is, well, "bizarre" is a rather kind characterization, and the actual mapping of `fcntl.lockf` (the only one which, from the documentation and other articles, seemed moderately sane) appears to be tied to `fd` rather than to the file system `inode`, which means, as I discovered in my quick experiment, it doesn't work at all between unrelated processes that don't share an `fd`. That is, I successfully executed `fcntl.lockf(lock, fcntl.LOCK_EX)` from two separate terminal windows on the same file. That's not helpful, and I'm going to have to recode this. Unfortunately, Python makes using the other file locking models even harder and less well defined, and I may have to try something completely different.

Reply:
A small experiment suggests that `LOCK_EX` is only available on files which are opened for Write access, and `LOCK_SH` is only available on files which are opened for Read access...so, if you want to convert from one to the other, the file has to be opened for _updating_ ([`r+`](https://docs.python.org/3/library/functions.html#open))...but, with that, I was able to freely convert back and forth between `EX` and `SH` (without going through `UN`) without incurring errors...of course, I don't know if the locking is actually _effective_, but it looks like we can convert....